src/main/scala/cognite/spark/v1/DefaultSource.scala

```diff
@@ -188,104 +188,108 @@ class DefaultSource
+          data.foreachPartition((rows: Iterator[Row]) => {
+            import CdpConnector.ioRuntime
+            val batches = rows.grouped(batchSize).toVector
+            batches.parTraverse_(relation.delete).unsafeRunSync()
+          })
+        } else {
+          // datapoints need special handling of dataframes and batches
+          relation.insert(data, overwrite = true)
+        }
+        relation
+      case relation: AssetHierarchyBuilder =>
+        config.onConflict match {
           case OnConflictOption.Delete =>
-            relation.delete
+            relation.delete(data)
+          case _ =>
+            relation.buildFromDf(data)
         }
+        relation
+      case relation: CdfRelation with WritableRelation =>
+        val batchSizeDefault = relation match {
+          case _: SequenceRowsRelation => Constants.DefaultSequenceRowsBatchSize
+          case _ => Constants.DefaultBatchSize
+        }
+        val batchSize = config.batchSize.getOrElse(batchSizeDefault)
+        val originalNumberOfPartitions = data.rdd.getNumPartitions
+        val idealNumberOfPartitions = config.sparkPartitions
+
+        // If we have very many partitions, it's quite likely that they are significantly uneven.
+        // And we will have to limit parallelism on each partition to low number, so the operation could
+        // take unnecessarily long time. Rather than risking this, we'll just repartition data in such case.
+        // If the number of partitions is reasonable, we avoid the data shuffling
+        val dataRepartitioned =
+          if (originalNumberOfPartitions > 50 && originalNumberOfPartitions > idealNumberOfPartitions) {
+            data.repartition(idealNumberOfPartitions)
+          } else {
+            data
+          }
 
-        batches
-          .parEvalMapUnordered(maxParallelism) { chunk =>
-            operation(chunk.toVector)
+        dataRepartitioned.foreachPartition((rows: Iterator[Row]) => {
+          import CdpConnector.ioRuntime
+
+          val maxParallelism = config.parallelismPerPartition
+          val batches = Stream.fromIterator[IO](rows, chunkSize = batchSize).chunks
+
+          val operation: Seq[Row] => IO[Unit] = config.onConflict match {
+            case OnConflictOption.Abort =>
+              relation.insert
+            case OnConflictOption.Upsert =>
+              relation.upsert
+            case OnConflictOption.Update =>
+              relation.update
+            case OnConflictOption.Delete =>
+              relation.delete
           }
-          .compile
-          .drain
-          .unsafeRunSync()
-      })
-      relation
+
+          batches
+            .parEvalMapUnordered(maxParallelism) { chunk =>
+              operation(chunk.toVector)
+            }
+            .compile
+            .drain
+            .unsafeRunSync()
+        })
+        relation
+      case _ =>
+        sys.error(s"Resource type $resourceType does not support save")
```

Comment:
```suggestion
        sys.error(s"Resource type $resourceType does not support save()")
```
maybe keep existing message